presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarArray.java

```diff
@@ -0,0 +1,158 @@
+
+    public static ColumnarArray toColumnarArray(Block block)
+    {
+        requireNonNull(block, "block is null");
+
+        if (block instanceof DictionaryBlock) {
+            return toColumnarArray((DictionaryBlock) block);
+        }
+        if (block instanceof RunLengthEncodedBlock) {
+            return toColumnarArray((RunLengthEncodedBlock) block);
+        }
+
+        if (!(block instanceof AbstractArrayBlock)) {
+            throw new IllegalArgumentException("Invalid row block");
+        }
+
+        AbstractArrayBlock arrayBlock = (AbstractArrayBlock) block;
+        Block elementsBlock = arrayBlock.getValues();
+
+        // trim elements to just visible region
+        int elementsOffset = 0;
+        int elementsLength = 0;
+        if (arrayBlock.getPositionCount() > 0) {
+            elementsOffset = arrayBlock.getOffset(0);
+            elementsLength = arrayBlock.getOffset(arrayBlock.getPositionCount()) - elementsOffset;
+        }
+        elementsBlock = elementsBlock.getRegion(elementsOffset, elementsLength);
+
+        return new ColumnarArray(block, arrayBlock.getOffsetBase(), arrayBlock.getOffsets(), elementsBlock);
+    }
+
+    private static ColumnarArray toColumnarArray(DictionaryBlock dictionaryBlock)
+    {
+        ColumnarArray columnarMap = toColumnarArray(dictionaryBlock.getDictionary());
+
+        // build new offsets
+        int[] offsets = new int[dictionaryBlock.getPositionCount() + 1];
+        for (int i = 1; i < offsets.length; i++) {
+            int dictionaryId = dictionaryBlock.getId(i - 1);
+            offsets[i] = offsets[i - 1] + columnarMap.getLength(dictionaryId);
+        }
+
+        // reindex dictionary
+        int[] dictionaryIds = new int[offsets[dictionaryBlock.getPositionCount()]];
+        int nextDictionaryIndex = 0;
+        for (int i = 1; i < offsets.length; i++) {
+            int dictionaryId = dictionaryBlock.getId(i - 1);
+            int length = columnarMap.getLength(dictionaryId);
+
+            int startOffset = columnarMap.getOffset(dictionaryId);
+            for (int entryIndex = 0; entryIndex < length; entryIndex++) {
+                dictionaryIds[nextDictionaryIndex] = startOffset + entryIndex;
+                nextDictionaryIndex++;
+            }
+        }
+
+        return new ColumnarArray(
+                dictionaryBlock,
+                0,
+                offsets,
+                new DictionaryBlock(dictionaryIds.length, columnarMap.getNullSuppressedElementsBlock(), dictionaryIds));
+    }
+
+    private static ColumnarArray toColumnarArray(RunLengthEncodedBlock rleBlock)
+    {
+        Block rleValue = rleBlock.getValue();
+
+        if (rleValue.isNull(0) || rleBlock.getPositionCount() == 0) {
+            return new ColumnarArray(
+                    rleBlock,
+                    0,
+                    new int[rleBlock.getPositionCount() + 1],
+                    new RunLengthEncodedBlock(rleValue, 0));
```

Comment:
I understand that this is an optimization, and is not required for correctness. But I'm curious as to why you have this here but not in row/map.